File: src/main/java/eu/yeger/minesweeper/Minesweeper.java
package eu.yeger.minesweeper;

import eu.yeger.minesweeper.controller.GameController;
import eu.yeger.minesweeper.model.Game;
import eu.yeger.minesweeper.model.DefaultModelBuilder;
import eu.yeger.minesweeper.model.ModelBuilder;
import eu.yeger.minesweeper.view.ViewBuilder;
import javafx.scene.Node;
import javafx.scene.image.Image;
import lombok.Builder;

@Builder
public class Minesweeper {

    @Builder.Default
    private int width = 10;
    @Builder.Default
    private int height = 10;
    @Builder.Default
    private int mineCount = 20;
    @Builder.Default
    private int cellSize = 40;

    private Image flagIcon;
    private Image mineIcon;

    private Runnable onGameWon;
    private Runnable onGameLost;

    private ModelBuilder modelBuilder;
    private ViewBuilder viewBuilder ;

    public Node instance() {
        defaultIconFallback();
        defaultBuilderFallback();
        final Game game = modelBuilder.build();
        new GameController(game, onGameWon, onGameLost);
        return viewBuilder.buildView(game);
    }

    private void defaultIconFallback() {
        if (flagIcon == null)
            flagIcon = new Image("/flag.png", cellSize, cellSize, true, true);
        if (mineIcon == null)
            mineIcon = new Image("/flag.png", cellSize, cellSize, true, true);
    }

    private void defaultBuilderFallback() {
        if (modelBuilder == null)
            modelBuilder = new DefaultModelBuilder(width, height, mineCount);
        if (viewBuilder == null)
            viewBuilder = new ViewBuilder(cellSize, flagIcon, mineIcon);
    }
}

File: src/main/java/eu/yeger/minesweeper/Minesweeper.kt
package eu.yeger.minesweeper

import eu.yeger.minesweeper.controller.GameController
import eu.yeger.minesweeper.model.DefaultModelBuilder
import eu.yeger.minesweeper.model.ModelBuilder
import eu.yeger.minesweeper.view.GameView
import javafx.scene.Node
import javafx.scene.image.Image

class Minesweeper {

    var width = 10
    var height = 10
    var mineCount = 20
    var cellSize = 40

    var flagIcon: Image? = null
    var mineIcon: Image? = null

    var onGameWon: (() -> Unit)? = null
    var onGameLost: (() -> Unit)? = null

    var modelBuilder: ModelBuilder? = null
    var gameView: GameView? = null

    fun instance(): Node {
        defaultIconFallback()
        defaultBuilderFallback()
        val game = modelBuilder!!.build()
        GameController(game, Runnable { onGameWon }, Runnable { onGameLost })
        return gameView!!.build(game).instance()
    }

    private fun defaultIconFallback() {
        if (flagIcon == null)
            flagIcon = Image("/flag.png", cellSize.toDouble(), cellSize.toDouble(), true, true)
        if (mineIcon == null)
            mineIcon = Image("/mine.png", cellSize.toDouble(), cellSize.toDouble(), true, true)
    }

    private fun defaultBuilderFallback() {
        if (modelBuilder == null)
            modelBuilder = DefaultModelBuilder(width, height, mineCount)
        if (gameView == null)
            gameView = GameView(cellSize, flagIcon!!, mineIcon!!)
    }
}
